### PR TITLE
Update Ember and Ember Data TS version to 3.7

### LINKED ISSUE #1152: Retype `assert` from `@ember/debug` as an assertion function
### Which package(s) does this problem pertain to?
  - [ ] @types/ember
  - [ ] @types/ember__string
  - [ ] @types/ember__polyfills
  - [ ] @types/ember__object
  - [ ] @types/ember__utils
  - [ ] @types/ember__array
  - [ ] @types/ember__engine
  - [ ] @types/ember__debug
  - [ ] @types/ember__runloop
  - [ ] @types/ember__error
  - [ ] @types/ember__controller
  - [ ] @types/ember__component
  - [ ] @types/ember__routing
  - [ ] @types/ember__application
  - [ ] @types/ember__test
  - [ ] @types/ember__test-helpers
  - [ ] @types/ember__service
  - [ ] @types/ember-data
  - [ ] @types/rsvp
  - [ ] Other
  - [ ] I don't know

### What are instructions we can follow to reproduce the issue?

```sh
ember new sample; cd ./sample # Create a new ember app
ember install ember-cli-typescript # Set up typescript support
```

##### Reproduction Case

```ts
import { assert } from '@ember/debug';

function foo(value: unknown): string {
  assert('value must be a string!', typeof value === 'string');
  return value.toUpperCase();
}
```

### Now about that bug. What did you expect to see?

Should pass type check

### What happened instead?

Does not pass type check

See https://github.com/ember-polyfills/ember-named-blocks-polyfill/blob/758619d0de8097bb8cdbebf5d54d3fadbfcc9295/addon/helpers/-ember-types.d.ts

### PR DESCRIPTION
**Pull request by chriskrycho**
This enables us to start adopting new features from more recent versions of TS, and is within our current official support commitments (latest two versions, even granting that these types were broken on 3.8).

Unfortunately, bumping the core types has a ripple effect, such that *every* type in the broader ecosystem which relies on them must *also* be updated. This  includes older versions specified by `typesVersions`, which *implicitly*  use the newer versions from the `@ember/*` and `@ember-data/*` types,  because of declaration merging.

Notable changes:

- `assert` now works with 3.7's `asserts` typing. This resolves typed-ember/ember-cli-typescript#1152 and it makes a *lot* of things *much* nicer. It's now possible to use Ember's built in `assert` like this:

    ```ts
    import { assert } from '@ember/debug';
    
    function dealsWithUnknown(val: unknown) {
      assert('this needs to be a string', typeof val === 'string');
      console.log(val.length); // type checks!
    }
    ```

- Closely related, there is a somewhat annoying workaround in the tests for microsoft/TypeScript#36931. It's probably not worth the Typed Ember team's time to actively document this, since to my knowledge no TS Ember user is currently using the namespace export, but this workaround is worth knowing about in case someone pops up and surprises us!

- We can now use `unknown` throughout. The only place I've used it so far is with the change to `assert` (I intended to keep this small, and only included the `assert` changes since they were the motivation for bumping everything to start with), but it'll be useful and more helpful for our consumers going forward. If and as we switch, however, we should make a point to publicize to the degree we can, since it's effectively a breaking change for our consumers.

- We can also now use some of the more interesting checks we'd long considered, but which were blocked on 2.9+ and 3.1+ in several cases.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

### CODE REVIEW

**Comment by chriskrycho** [comment]
I'll dig in Monday morning to fix the issues highlighted by CI!

File: types/ember__utils/index.d.ts
Comment (chriskrycho) [inline types/ember__utils/index.d.ts:57]: Note for myself on Monday: revert this! it wasn't intended to make it in; auto-formatting got me.

File: types/ember/test/debug.ts
Comment (chriskrycho) [inline types/ember/test/debug.ts:13]: Note for myself on Monday: this isn't a complete sentence. 😂

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ember-feature-flags ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-feature-flags@latest/))
was missing the following properties:
1. name
1. name
## ember-mocha ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-mocha@latest/))
was missing the following properties:
1. name
1. init
1. postBuild
1. checkPackages
1. included

<details>
<summary>as well as these 17 other properties...</summary>
<p>targetOptions, contentFor, treeForVendor, treeForAddonTestSupport, overrideTestCommandFilter, setTestGenerator, name, init, postBuild, checkPackages, included, targetOptions, contentFor, treeForVendor, treeForAddonTestSupport, overrideTestCommandFilter, setTestGenerator</p>
</details>

## ember-qunit ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-qunit@latest/))
was missing the following properties:
1. name
1. init
1. included
1. targetOptions
1. contentFor

<details>
<summary>as well as these 11 other properties...</summary>
<p>treeForVendor, treeForAddonTestSupport, setTestGenerator, name, init, included, targetOptions, contentFor, treeForVendor, treeForAddonTestSupport, setTestGenerator</p>
</details>

## ember-resolver ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-resolver@latest/))
was missing the following properties:
1. name
1. emberResolverFeatureFlags
1. init
1. treeForAddon
1. included

<details>
<summary>as well as these 7 other properties...</summary>
<p>monkeyPatchVendorFiles, name, emberResolverFeatureFlags, init, treeForAddon, included, monkeyPatchVendorFiles</p>
</details>

## ember__test-helpers ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember__test-helpers@latest/))
was missing the following properties:
1. name
1. init
1. included
1. treeForAddonTestSupport
1. treeForVendor

<details>
<summary>as well as these 20 other properties...</summary>
<p>name, init, included, treeForAddonTestSupport, treeForVendor, name, init, included, treeForAddonTestSupport, treeForVendor, name, init, included, treeForAddonTestSupport, treeForVendor, name, init, included, treeForAddonTestSupport, treeForVendor</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e83c113b8542543e1e6f622670bd2236cc1876c2
</p>

File: types/ember-data__adapter/index.d.ts
Comment (dfreeman) [inline types/ember-data__adapter/index.d.ts:1]: If we walk the version _back_ for the `@ember-data/*` packages, even if the `latest` tag points to 3.12.x, isn't there a danger of folks winding up with an older `3.15.x` version in a some cases?

File: types/ember/test/debug.ts
Comment (dfreeman) [inline types/ember/test/debug.ts:15]: I think this will also work (since the rule is really _just_ "there must be a type annotation"):

```ts
const assert: typeof Ember.assert = Ember.assert;
```

https://www.typescriptlang.org/play/index.html?ssl=6&ssc=1&pln=7&pc=1#code/CYUwxgNghgTiAEA7KBbEBnADlMCCiKARiDPAN4BQ818AZgK6JgAuAlgPaLxTronMAKUOjAAueOmYxWiAOYBKcYhAA3EgG4qNBkzaduvfkIxiJUmbIA08ZhmbjGAa0TsA7okUG+MZuht3NAF8KEIowTkkvfnFmAE9MEHZaeAJiGAA6Hm9meABeFKISTMMfTVDQSFgECBAcgA9xSWk5eAAfJHoICE0sowByOvhWPyl6ZgALWL7rOvkyuvTmdgBVTASYAGEeEAE5+AB6fe4u+Fl2dmB1eFcEXp8QYHhB4ZsYMcnQu8E+wnOUAEI+nMwhF2DV0hB2LIBH0Ji8Xow4DhxlBCDV4OFQNxEI90ON2J1HsR4ONWLJxhAyeNbI8eBJ6GBxkDNEA

Comment (chriskrycho) [reply types/ember/test/debug.ts]: Ah, very good. (I thought I had tried this, but I think I actually tried using a cast `as typeof Ember.assert` instead.)

File: types/ember-data__adapter/index.d.ts
Comment (chriskrycho) [reply types/ember-data__adapter/index.d.ts]: Good point. Accordingly, and given 3.12 is in security-fixes-only LTS, I went ahead and bumped to 3.16 for all the core Ember and Ember Data types!

**Review by dfreeman — APPROVED** [review]
Looks good to me! Excited for some of the possibilities this opens up for cleanup in the types 🙂

**Review by jamescdavis — APPROVED** [review]
This is :100:!

**Comment by orta** [comment]
I've started the re-running of the CI which failed (it took 6 hours last time and timed out... which err, might mean I choose to ignore it if it happens again ) 

Other than that, I'm happy we've got enough approvals, so will merge after that

**Comment by chriskrycho** [comment]
Thanks for the update! No surprise it took a long time with this many impacted packages. :sweat_smile:

**Comment by raycohen** [comment]
the type for `assert(string)` doesn't understand that asserts are removed from ember production builds. This is giving us typescript unreachable code warnings for code that will be reachable once the production build removes assert. We can override this definition in our app to allow our usage, but I'm wondering if this would be useful for these types to know

**Comment by chriskrycho** [comment]
@raycohen that’s because assert in that form represents a hard error. Even though it won’t run the same way in development and production, that is equally true of using assertions for type-narrowing, so by the same reasoning, you should have to do additional work for the other forms of assert as well. You should only assert in debug where you are treating it as a genuinely hard failure, and choosing to leave that out for speed reasons for your production build.

In sum: TS and these types are correct to test it as dead code; doing anything else would make asserts useless for type info.

**Pull request merged**
